packages/gatsby-plugin-cms/src/node-api/sourceNode.ts

```diff
@@ -0,0 +1,83 @@
+import camelcase from 'camelcase'
+import type { ParentSpanPluginArgs } from 'gatsby'
+
+import { PLUGIN } from '../constants'
+import type { RemotePageContent, Block, PageContent } from './types'
+
+export const getTypeName = (name: string) => camelcase(['cms', name])
+
+const baseSchema = `
+scalar JSONPropsCmsObject
+
+type CmsBlock {
+  name: String!
+  props: JSONPropsCmsObject!
+}
+`
+
+export const createSchemaCustomization = (
+  gatsbyApi: ParentSpanPluginArgs,
+  nodes: RemotePageContent[]
+) => {
+  const getType = (node: RemotePageContent) => `
+    type ${getTypeName(node.type)} implements Node {
+      id: String!
+      name: String!
+      sections: [CmsBlock!]!
+    }
+    `
+
+  const types = Array.from(new Set(nodes.map(getType)))
+
+  const typeDefs = [baseSchema, ...types].join('\n')
+
+  gatsbyApi.actions.createTypes(typeDefs)
+}
+
+const nodeId = (node: RemotePageContent): string =>
+  `${getTypeName(node.type)}:${node.remoteId}`
+
+export const sourceNode = (
+  gatsbyApi: ParentSpanPluginArgs,
+  node: RemotePageContent
+) => {
+  const extra = node.extraBlocks.reduce(
+    (acc, { name, blocks }) => ({
+      ...acc,
+      [camelcase(name)]: blocks.reduce(
+        (ac, block) => ({ ...ac, [camelcase(block.name)]: block.props }),
+        {} as Record<string, Block['props']>
+      ),
+    }),
+    {} as Record<string, Record<string, Block['props']>>
+  )
```

Comment:
What do you think to change this reduce to a forEach, it will make it more clear. Another point, is that this algorithm is doing a DFS, why not convert it to a dfs algorithm?

```suggestion
  const extra = traverseExtraBlocks(node.extraBlocks)
```

```ts
type BlockTree = Record<string, Block['props']> 

const traverseExtraBlocks = (blockNodes: typeBlockHere): BlockTree<TypeBlockHere> => {
   const extraBlockTree = {}
   blockNodes.forEach((block) => {
// fix the algorithm here if needed
     if (Array.isArray(block.blocks) {
        extraBlockTree[camelcase(block.name)] = traverseExtraBlocks(block.blocks)
     }
  })
  return extraBlockTree
}
```

Reply:
I don't think so. This is not a dfs, but rather a two level iteration on the object

Reply:
Ok, nice! What do you think to transform the reduce into a forEach?

Reply:
I'm not sure. Are you concerned with memory usage? `forEach` is more performatic but reduce a declarative construct.